zookeeper-api/src/main/java/org/apache/helix/zookeeper/routing/ZkRoutingDataReader.java

```diff
@@ -0,0 +1,83 @@
+package org.apache.helix.zookeeper.routing;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.helix.msdcommon.constant.MetadataStoreRoutingConstants;
+import org.apache.helix.zookeeper.datamodel.ZNRecord;
+import org.apache.helix.zookeeper.datamodel.serializer.ZNRecordSerializer;
+import org.apache.helix.zookeeper.exception.MultiZkException;
+import org.apache.helix.zookeeper.impl.client.ZkClient;
+import org.apache.helix.zookeeper.zkclient.exception.ZkNoNodeException;
+
+
+/**
+ * Zk-based RoutingDataReader that establishes a ZK connection to the routing ZK to fetch routing
+ * data.
+ * The reading of routing data by nature should only be performed in cases of a Helix client
+ * initialization or routing data reset. That means we do not have to maintain an active ZK
+ * connection. To minimize the number of client-side ZK connections, ZkRoutingDataReader establishes
+ * a ZK session temporarily only to read from ZK afresh and closes sessions upon read completion.
+ */
+public class ZkRoutingDataReader implements RoutingDataReader {
+
+  /**
+   * Returns a map form of metadata store routing data.
+   * The map fields stand for metadata store realm address (key), and a corresponding list of ZK
+   * path sharding keys (key).
+   * @param endpoint
+   * @return
+   */
+  @Override
+  public Map<String, List<String>> getRawRoutingData(String endpoint) {
+    ZkClient zkClient =
+        new ZkClient.Builder().setZkServer(endpoint).setZkSerializer(new ZNRecordSerializer())
+            .build();
```

Comment:
Do we want to move this into try? Is that possible any exception happening in build zkClient?

Reply:
If this throws an exception, we should let it throw the exception and not proceed. We should fail early.